neqo-http3/src/server.rs

```diff
@@ -111,31 +113,59 @@ impl Http3Server {
                 .borrow_mut()
                 .process_http3(&mut conn.borrow_mut(), now);
             let mut remove = false;
-            while let Some(e) = handler.borrow_mut().next_event() {
-                match e {
-                    Http3ServerConnEvent::Headers {
-                        stream_id,
-                        headers,
-                        fin,
-                    } => self.events.headers(
-                        ClientRequestStream::new(conn.clone(), handler.clone(), stream_id),
-                        headers,
-                        fin,
-                    ),
-                    Http3ServerConnEvent::Data {
-                        stream_id,
-                        data,
-                        fin,
-                    } => self.events.data(
-                        ClientRequestStream::new(conn.clone(), handler.clone(), stream_id),
-                        data,
-                        fin,
-                    ),
-                    Http3ServerConnEvent::StateChange(state) => {
-                        self.events
-                            .connection_state_change(conn.clone(), state.clone());
-                        if let Http3State::Closed { .. } = state {
-                            remove = true;
+            {
+                let mut handler_borrowed = handler.borrow_mut();
+                while let Some(e) = handler_borrowed.next_event() {
+                    match e {
+                        Http3ServerConnEvent::Headers {
+                            stream_id,
+                            headers,
+                            fin,
+                        } => self.events.headers(
+                            ClientRequestStream::new(conn.clone(), handler.clone(), stream_id),
+                            headers,
+                            fin,
+                        ),
+                        Http3ServerConnEvent::DataReadable { stream_id } => {
+                            loop {
```

Comment:
I realize that Rust might not let you do this, but this function is an epic.  Please consider moving this loop to a new function.